domain/robstoll-lib/atrium-domain-robstoll-lib-jvm/src/main/kotlin/ch/tutteli/atrium/domain/robstoll/lib/creating/filesystem/symbolicLinkResolving.kt

```
@@ -0,0 +1,101 @@
+    visitedList: MutableList<Path>,
+    initialPath: Path
+): Path {
+    var currentPath = absolutePath.root
+    visitedList.add(absolutePath)
+
+    for (part in absolutePath) {
+        currentPath = currentPath.resolve(part)
+        val nextPath = addOneStepResolvedSymlinkHint(
+            currentPath,
+            hintList
+        )
+        if (nextPath != null) {
+            val visitedIndex = visitedList.indexOf(nextPath)
+            if (visitedIndex != -1) {
+                repeat(hintList.size - visitedIndex) { hintList.removeLast() }
+                // add to the list so [hintForLinkLoop] prints this duplicate twice
+                visitedList.add(nextPath)
+                hintList.add(hintForLinkLoop(visitedList, visitedIndex))
+                return initialPath
+            } else {
+                currentPath = addAllLevelResolvedSymlinkHints(nextPath, hintList, visitedList, initialPath)
+            }
+        }
+    }
+    return currentPath
+}
+
+/**
+ * If [absolutePath] is surely a symlink, adds an explanatory hint to [resultList] and returns the link target.
+ * Return `null` and does not modify [resultList] otherwise.
+ */
+private fun addOneStepResolvedSymlinkHint(absolutePath: Path, resultList: Deque<Assertion>): Path? {
+    try {
+        val nextPath = Files.readSymbolicLink(absolutePath)
+        resultList.add(
+            ExpectImpl.builder.explanatory
+                .withExplanation(HINT_FOLLOWED_SYMBOLIC_LINK, absolutePath, nextPath)
+                .build()
+        )
+        return nextPath
+    } catch (e: IOException) {
+        // either this is not a link, or we cannot check it. The best we can do is assume it is not a link.
+        return null
+    }
+}
+
+private fun hintForLinkLoop(loop: List<Path>, startIndex: Int): Assertion {
+    val loopRepresentation = loop.subList(startIndex, loop.size).joinToString(" -> ")
+    return ExpectImpl.builder.explanatoryGroup.withWarningType.withAssertion(
```

you could also use `withExplanatoryAssertion` instead of `withAssertion` and use a placeholder in `FAILURE_DUE_TO_LINK_LOOP` to pass the loopRepresentation

---

Good idea! Done.